pkg/kubelet/cm/cpumanager/policy_static.go

```diff
@@ -354,6 +354,74 @@ func (p *staticPolicy) GetTopologyHints(s state.State, pod *v1.Pod, container *v
 	}
 }
 
+func (p *staticPolicy) getPodRequestedCPU(pod *v1.Pod) int {
+	// The maximum of requested CPUs by init containers.
+	requestedByInitContainers := 0
+	for _, container := range pod.Spec.InitContainers {
+		if _, ok := container.Resources.Requests[v1.ResourceCPU]; !ok {
+			continue
+		}
+		requestedCPU := p.guaranteedCPUs(pod, &container)
+		if requestedCPU > requestedByInitContainers {
+			requestedByInitContainers = requestedCPU
+		}
+	}
+	// The sum of requested CPUs by app containers.
+	requestedByAppContainers := 0
+	for _, container := range pod.Spec.Containers {
+		if _, ok := container.Resources.Requests[v1.ResourceCPU]; !ok {
+			continue
+		}
+		requestedByAppContainers += p.guaranteedCPUs(pod, &container)
+	}
+
+	requestedByPod := requestedByAppContainers
+	if requestedByInitContainers > requestedByAppContainers {
+		requestedByPod = requestedByInitContainers
+	}
+	klog.Errorf("[cpumanager] requestedByInits : %d, requestedByUsers : %d, requestedByPod : %d", requestedByInitContainers, requestedByAppContainers, requestedByPod)
+	return requestedByPod
+}
+
+func (p *staticPolicy) GetPodLevelTopologyHints(s state.State, pod *v1.Pod) map[string][]topologymanager.TopologyHint {
+	// Get a count of how many guaranteed CPUs have been requested by Pod.
+	requestedByPod := p.getPodRequestedCPU(pod)
+
+	if requestedByPod == 0 {
+		return nil
```

Comment:
Does this signify an error? Would it mean non-admittance? Why it's nil here and empty map in case of error below?

Reply:
It's not an error. If pod is not a Guaranteed QoS class pod `requestedByPod` will be equal to `0`, because function returns CPUs only when pod has that QoS class. 

returning `nil` will be treated as providing hint `NUMANodeAffinity: nil, Preferred: false`, which can mean non-admission but it depends on the Topology Manager policy

Reply:
I can confirm that `nil` is what you want to return here. However, the interpretation of `nil` as `NUMANodeAffinity: nil, Preferred: false` is incorrect.

The topologymanager will actually treat this as a `NUMANodeAffinity: nil, Preferred: true` as shown here:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/topologymanager/policy.go#L71

This essentially translates to a `no preference`, which causes the topologymanager to ignore this resource as something that should be considered for alignment.

Reply:
Yes, that is right, I even checked that before answering, however, I mixed it in the answer. Thanks for correcting

Reply:
would it benefit to leave a comment here? It's not easy to understand by looking at method code how `nil` is different from the empty collection below

Reply:
At the hint's perspective, `nil` means no-preference. So, pod does not have prefer resource and any resource can be allocated to the pod. But, in case of empty list `{ }`, the pod has preference and there is no resource that satisfies the preferences that pod wants.

Reply:
I'd suggest to add comment next to return statement or in the description of the whole method what every return means. The knowledge that `nil` is so much different that empty collection is non-trivial and need to be explicit.

Reply:
Something like this for above code part is understandable and enough explanatory?
```
// Number of required CPU is not integer or a pod hasn't got a Guaranteed QoS class.
// It will be treated by Topology Manager as having no preference and cause to ignore this
// resource while considering a pod alignment.
// In definition of hints this is equal to: TopologyHints[NUMANodeAffinity: nil, Preferred: true].
	if requestedByPod == 0 {
		return nil
	}
```

Reply:
sounds great for me. I think having the comment will help to maintain code long term

Reply:
Comment applied.